File: dsa/04-queue/queueLinkList.cpp
#include <iostream>
#include <iostream>

using std::cout;
using std::endl;

class ListNode
{
public:
    int val_;
    ListNode *next = nullptr;

    ListNode(int val)
    {
        val_ = val;
    }
};

class Queue
{
public:

    ListNode *front = nullptr;
    ListNode *rear = nullptr;

    Queue() {}

    void enqueue(int val)
    {
        ListNode *newNode = new ListNode(val);

        // Queue is non-empty
        if (rear != nullptr)
        {
            rear->next = newNode;
            rear = rear->next;
        }
        // Queue
        else
        {
            front = rear = newNode;
        }
    }

    int dequeue()
    {
        // Queue is empty
        if (front == nullptr)
        {
            return -1; // Better to throw an exception
        }

        int val = front->val_;
        front = front->next;
        return val;
    }

    void print()
    {
        ListNode *curr = front;
        while (curr != nullptr)
        {
            cout << curr->val_ << " -> ";
            curr = curr->next;
        }
        cout << endl;
    }
};

int main()
{
    Queue q = Queue();
    q.enqueue(12);
    q.enqueue(10);
    cout << q.dequeue() << endl;
    q.print();
    return 0;
}

File: dsa/04-queue/queueLinkList.py
class ListNode:
    def __init__(self, data: int) -> None:
        self.data = data
        self.next = None


class Queue:
    def __init__(self) -> None:
        self.front = self.rear = None

    def enQueue(self, data: int):
        newNode = ListNode(data)

        # rear not None
        if self.rear:
            self.rear.next = newNode
            self.rear = None # self.rear = self.rear.next

        else:
            self.front = self.rear = newNode

    def deQueue(self):
        if self.front is None:
            return None
        else:
            val=self.front.data
            self.front=self.front.next
            return val
    def traverse(self):
        curr=self.front
        while curr:
            print(curr.data, ' -> ', end ="")
            curr = curr.next

if __name__=='__main__':
    q=Queue()
    q.enQueue(12)
    q.enQueue(10)
    print(q.deQueue())
    q.traverse()